### PR TITLE
Profitbricks REST - Volume API

### PR DESCRIPTION
**Pull request by mirza-spc**
@jasminSPC

### CODE REVIEW

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/volume/CreateVolumeRequestBinder.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+         properties.put("imagePassword", payload.imagePassword());
+      
+      if (payload.image() != null)
+         properties.put("image", payload.image());
+      else if (payload.licenceType() != null)
+         properties.put("licenceType", payload.licenceType());
+      
+      requestBuilder.put("properties", properties);
+      
+      return (new Gson()).toJson(requestBuilder);
```
Comment (devcsrj) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/volume/CreateVolumeRequestBinder.java]: Use the `Json` object, as you did with [here](https://github.com/jclouds/jclouds-labs/blob/master/profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/server/AttachCdromRequestBinder.java#L52).

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/volume/UpdateVolumeRequestBinder.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      if (payload.name() != null)
+         requestBuilder.put("name",  payload.name());
+      
+      if (payload.size() != null)
+         requestBuilder.put("size",  payload.size());
+      
+      if (payload.bus() != null)
+         requestBuilder.put("bus",  payload.bus());
+      
+      return (new Gson()).toJson(requestBuilder);
```
Comment (devcsrj) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder/volume/UpdateVolumeRequestBinder.java]: Use `Json` here too.

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/VolumeApi.java [hunk 3]
```diff
@@ -19,20 +19,118 @@
...
+   @Named("volume:snapshot:create")
+   @POST
+   @MapBinder(CreateSnapshotRequestBinder.class)
+   @ResponseParser(SnapshotApi.SnapshotParser.class)
+   Snapshot createSnapshot(@PayloadParam("snapshot") Volume.Request.CreateSnapshotPayload payload);
+   
+   @Named("volume:snapshot:restore")
+   @POST
+   @MapBinder(RestoreSnapshotRequestBinder.class)
+   @Fallback(Fallbacks.VoidOnNotFoundOr404.class)
```
Comment (devcsrj) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/VolumeApi.java]: Remove fallback. Let's let the user know why the *restore* op failed

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/ServerApiMockTest.java [hunk 4]
```diff
@@ -65,7 +65,7 @@ public void testGetListWithDepth() throws InterruptedException {
    
    @Test
    public void testGetListWith404() throws InterruptedException {
```
Comment (devcsrj) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/ServerApiMockTest.java]: You may remove the test for 404 fallback, as this is already tested in the core.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+import org.apache.jclouds.profitbricks.rest.domain.LicenceType;
+import org.apache.jclouds.profitbricks.rest.domain.Volume;
+import org.apache.jclouds.profitbricks.rest.domain.options.DepthOptions;
+import org.apache.jclouds.profitbricks.rest.internal.BaseProfitBricksApiMockTest;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "VolumeApiMockTest", singleThreaded = true)
```
Comment (devcsrj) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java]: Hmm. Do you need this test to be `singleThreaded`?

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+      assertNotNull(list);
+      assertEquals(list.size(), 6);
+      assertEquals(list.get(0).properties().name(), "Docker Registry Volume");
+      
+      assertEquals(server.getRequestCount(), 1);
+      assertSent(server, "GET", "/datacenters/datacenter-id/volumes");
+   }
+   
+   @Test
+   public void testGetListWith404() throws InterruptedException {
```
Comment (devcsrj) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java]: No need to test this (already expected to return an empty list, because of the declared, already-tested fallback)

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+      Volume volume = volumeApi().getVolume("datacenter-id", "some-id");
+      
+      assertNotNull(volume);
+      assertEquals(volume.properties().name(), "Docker Registry Volume");
+      
+      assertEquals(server.getRequestCount(), 1);
+      assertSent(server, "GET", "/datacenters/datacenter-id/volumes/some-id");
+   }
+   
+   public void testGetVolumeWith404() throws InterruptedException {
```
Comment (devcsrj) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java]: No need for this too.

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+         new MockResponse().setBody("")
+      );
+      
+      volumeApi().deleteVolume("datacenter-id", "some-id");
+      assertEquals(server.getRequestCount(), 1);
+      assertSent(server, "DELETE", "/datacenters/datacenter-id/volumes/some-id");
+   }
+   
+   @Test
+   public void testDeleteWith404() throws InterruptedException {
```
Comment (devcsrj) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java]: Same as above

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/internal/BaseProfitBricksApiMockTest.java [hunk 9]
```diff
@@ -89,13 +89,13 @@ protected MockResponse response204() {
    protected MockResponse response404() {
```
Comment (devcsrj) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/internal/BaseProfitBricksApiMockTest.java]: No need for this then. :) Also, remove 404 tests (or Unauthorized, if exists) from other tests, if possible.

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/util/MacAddresses.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+package org.apache.jclouds.profitbricks.rest.util;
+
+import java.util.regex.Pattern;
+
+public class MacAddresses {
+
+   private static final String MAC_ADDR_FORMAT = "^([0-9a-f]{2}[:]){5}([0-9a-f]{2})$";
+   private static final Pattern MAC_ADDR_PATTERN = Pattern.compile(MAC_ADDR_FORMAT);
+
+   public static boolean isMacAddress(String in) {
```
Comment (devcsrj) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/util/MacAddresses.java]: Add missing test

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/util/Passwords.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+ * limitations under the License.
+ */
+package org.apache.jclouds.profitbricks.rest.util;
+
+import java.util.Random;
+import java.util.regex.Pattern;
+
+import com.google.common.collect.ImmutableSet;
+
+public class Passwords {
```
Comment (devcsrj) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/util/Passwords.java]: Add test for this too, so we can validate it's [usage](https://github.com/jclouds/jclouds-labs/pull/233/files#diff-6d0cd6b1064677eca969d404ba5c0a91R175).

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/VolumeApi.java [hunk 12]
```diff
@@ -19,20 +19,118 @@
...
+   @DELETE
+   @Path("/{volumeId}")
+   @Fallback(Fallbacks.VoidOnNotFoundOr404.class)
+   void deleteVolume(@PathParam("dataCenterId") String dataCenterId, @PathParam("volumeId") String volumeId);
+   
+   @Named("volume:snapshot:create")
+   @POST
+   @MapBinder(CreateSnapshotRequestBinder.class)
+   @ResponseParser(SnapshotApi.SnapshotParser.class)
+   Snapshot createSnapshot(@PayloadParam("snapshot") Volume.Request.CreateSnapshotPayload payload);
```
Comment (devcsrj) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/VolumeApi.java]: Hmm. Should this (and `restoreSnapshot`) be in `SnapshotApi.java` instead? If so, don't forget to move both of their [tests](https://github.com/jclouds/jclouds-labs/pull/233/files#diff-81aadea0059ec0e66cf6304f6234762bR148) too!

**Comment by devcsrj** [comment]
Hi @mirza-spc !

Did initial review for the PR, while waiting for others to jump in. :) In addition to the comments above, could you add a test for [`ParseId`](https://github.com/StackPointCloud/jclouds-labs/commit/60869d038e8022165b6b1c32cb80233cee213033#diff-640a66e9fc832b6c994d936877e933caR29)? This way, we're confident that the `#parseId` method actually returns what we expect. And, can you move the [`Pattern.compile(..)`](https://github.com/StackPointCloud/jclouds-labs/commit/60869d038e8022165b6b1c32cb80233cee213033#diff-640a66e9fc832b6c994d936877e933caR43) as a class level [property](https://github.com/StackPointCloud/jclouds-labs/commit/60869d038e8022165b6b1c32cb80233cee213033#diff-640a66e9fc832b6c994d936877e933caR32), so that the pattern is only compiled once (and not everytime the method is called).

File: profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java (hunk 5 above)
Comment (mirza-spc) [inline profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/features/VolumeApiMockTest.java]: yes, because of server.enqueue

File: profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/VolumeApi.java (hunk 12 above)
Comment (mirza-spc) [inline profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/VolumeApi.java]: I can see your point, but then again a Volume is needed as well. Also it's under volumes at https://devops.profitbricks.com/api/rest/#volumes

**Comment by mirza-spc** [comment]
`Pattern.compile(String.format("..%s..", prefix))` uses a method param "prefix", so I don't think that I could move it to a static variable. This is unless there is some way to do this, of which I'm not aware.

**Comment by mirza-spc** [comment]
Also since ParseId test would require injected ParseId (because of Json), I'm wondering how I can achieve that.

**Comment by nacx** [comment]
> Also since ParseId test would require injected ParseId (because of Json), I'm wondering how I can achieve that.

You can create a small Guice context, with the json configuration, for that test:

```java
Injector injector = Guice.createInjector(new GsonModule());
ParseId parseId = injector.getInstance(ParseId.class);
```

**Comment by nacx** [comment]
Thanks for the fixes @mirza-spc!
@devcsrj Wanna review the fixes before squashing the commits?

**Comment by devcsrj** [comment]
Thanks for the `injector`-approach @nacx !

@mirza-spc I noticed that there are missing test in the [`-Binder`](https://github.com/StackPointCloud/jclouds-labs/tree/profitbricks-volume-api/profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/binder) classes. Could you add tests for these so that we can verify as to whether those binders actually build the correct payload? (I for one usually *trip* on these types, like typos in property names, etc. :) ) After that, this PR looks good to me!

**Comment by mirza-spc** [comment]
Seems like the tests are failing because of the different ordering of keys in jsonBinder.toJson method. Any thoughts on this?

**Comment by nacx** [comment]
You can start from an object for the expected payload and serialize it to get the expected payload string. This way you should be able to avoid ordering issues caused by different JVM defaults.

**Comment by devcsrj** [comment]
Binder tests, :heavy_check_mark: ! Unless @nacx has further comments, this LGTM!

**Comment by nacx** [comment]
@mirza-spc mind squashing the commits?

**Comment by mirza-spc** [comment]
So I merged the upstream. However if I squash it as well, the conflict with the upstream will arise again. So I've left it be for now.

**Comment by devcsrj** [comment]
Hmm. Unfortunately, this branch needs to be rebased and squashed, such that it contains one commit with changes only related to this PR. You'll notice that old files, such as [this](https://github.com/jclouds/jclouds-labs/pull/233/files#diff-5f712604f84e8ddc5e1fb9ca7f759697R51) no longer exists in `jclouds/jclouds-labs`, but is existing here. 

I'll try to rebase later so that this gets merged, and we can proceed to #241 .

**Comment by devcsrj** [comment]
Pushed to master as [9a8b346](https://github.com/jclouds/jclouds-labs/commit/9a8b346cafd9e332b0e532f1341e739d0388a347). Thanks for the effort, @mirza-spc !

**Comment by nacx** [comment]
Just noticed this while reviewing the last commit. Is the [commons-lang](https://github.com/jclouds/jclouds-labs/blob/master/profitbricks-rest/pom.xml#L68-L74) dependency really needed?

**Comment by nacx** [comment]
And even if it is needed we can't rely on a snapshot dependency, to it must be removed.

**Comment by devcsrj** [comment]
Woops, missed that. Pushed a [commit](https://github.com/jclouds/jclouds-labs/commit/5bd8447b28fe8b0b7a2b85b03270de643a6a62c0) to remove the dep.

**Comment by jasminSPC** [comment]
Is this completed? Can we rebase #241?

**Comment by devcsrj** [comment]
Yes! This has been pushed to master. You may now rebase the other PR.
